zerver/lib/actions.py

```diff
@@ -4845,70 +4846,178 @@ def do_change_can_create_users(user_profile: UserProfile, value: bool) -> None:
+                event_time=timezone_now(),
+                extra_data=orjson.dumps(
+                    {
+                        RealmAuditLog.OLD_VALUE: old_is_web_public_value,
+                        RealmAuditLog.NEW_VALUE: stream.is_web_public,
+                    }
+                ).decode(),
+            )
 
-    # We reuse "invite_only" stream update API route here because
-    # both are similar events and similar UI updates will be required
-    # by the client to update this property for the user.
     event = dict(
         op="update",
         type="stream",
         property="invite_only",
-        value=False,
-        history_public_to_subscribers=True,
-        is_web_public=True,
+        value=stream.invite_only,
+        history_public_to_subscribers=stream.history_public_to_subscribers,
+        is_web_public=stream.is_web_public,
         stream_id=stream.id,
         name=stream.name,
     )
     send_event(stream.realm, event, can_access_stream_user_ids(stream))
 
+    old_policy_name = get_stream_permission_policy_name(
+        old_invite_only_value, old_history_public_to_subscribers_value, old_is_web_public_value
+    )
+    new_policy_name = get_stream_permission_policy_name(
+        stream.invite_only, stream.history_public_to_subscribers, stream.is_web_public
+    )
+    send_change_stream_permission_notification(
+        acting_user, stream, old_policy_name, new_policy_name
+    )
 
-def do_change_stream_permission(
-    stream: Stream,
-    invite_only: Optional[bool] = None,
-    history_public_to_subscribers: Optional[bool] = None,
-    is_web_public: Optional[bool] = None,
+
+def send_change_stream_post_policy_notification(
+    stream: Stream, post_policy: int, user_profile: UserProfile
 ) -> None:
-    # TODO: Ideally this would be just merged with do_change_stream_invite_only.
-    if is_web_public:
-        do_make_stream_web_public(stream)
+    sender = get_system_bot(settings.NOTIFICATION_BOT, user_profile.realm_id)
+    user_mention = silent_mention_syntax_for_user(user_profile)
+
+    if post_policy == Stream.STREAM_POST_POLICY_EVERYONE:
+        post_policy_name = gettext_lazy("All stream members can post")
+    elif post_policy == Stream.STREAM_POST_POLICY_ADMINS:
+        post_policy_name = gettext_lazy("Only organization administrators can post")
+    elif post_policy == Stream.STREAM_POST_POLICY_MODERATORS:
+        post_policy_name = gettext_lazy("Only organization administrators and moderators can post")
     else:
-        assert invite_only is not None
-        do_change_stream_invite_only(stream, invite_only, history_public_to_subscribers)
+        post_policy_name = gettext_lazy("Only organization full members can post")
+
+    with override_language(stream.realm.default_language):
+        notification_string = _(
+            "{user} changed the [posting permissions](/help/stream-sending-policy) "
+            "for this stream. {post_policy}."
```

Comment:
This seems like an odd phrasing, in that I think it won't be clear that we're explaining the post-change policy.  @alya FYI to help draft a revision.

Once that's done, can you post screenshots of sample messages from each of these in #design for discussion?